src/utils/PathUtils.ts

```diff
@@ -0,0 +1,127 @@
+    pathPart.brackets?.forEach(bracket => (path += `[${bracket}]`));
+    if (i < pathParts.length - 1) {
+      path += '.';
+    }
+  });
+  return path;
+}
+
+/**
+ * Replaces soft indexs in rule paths with corresponding numbers
+ * @param {Rule[]} rules - An array of Rules
+ */
+export function resolveSoftIndexing(rules: Rule[]): void {
+  const pathMap: Map<string, number> = new Map();
+  const caretPathMap: Map<string, Map<string, number>> = new Map();
+
+  // Parsing and separating rules by base name and bracket indexes
+  const parsedRules = rules.map(rule => {
+    const parsedPath: any = { path: parseFSHPath(rule.path) };
+    // If we have a CaretValueRule, we'll need a second round of parsing for the caret path
+    if (rule instanceof CaretValueRule) {
+      parsedPath.caretPath = parseFSHPath(rule.caretPath);
+    }
+    return parsedPath;
+  });
+
+  // Replacing Soft indexes with numbers
+  parsedRules.map(rule => {
+    rule.path.forEach((element: PathPart) => {
+      if (!pathMap.has(element.base)) {
+        pathMap.set(element.base, 0);
+        if (element.brackets?.includes('+')) {
+          element.brackets[element.brackets.indexOf('+')] = '0';
+        }
+      } else {
+        element.brackets?.forEach((bracket: string, index: number) => {
+          if (bracket === '+') {
+            const newIndex = pathMap.get(element.base) + 1;
+            element.brackets[index] = newIndex.toString();
+            pathMap.set(element.base, newIndex);
+          } else if (bracket === '=') {
+            const currentIndex = pathMap.get(element.base);
+            element.brackets[index] = currentIndex.toString();
+          }
+        });
+      }
+    });
+    rule.path = assembleFSHPath(rule.path); // Assembling the separated rule path back into a normal string
+
+    rule.caretPath?.forEach((element: PathPart) => {
+      // Caret path indexes should only be resolved in the context of a specific path
+      // Each normal path has a separate map to keep track of the caret path indexes
+      if (!caretPathMap.has(rule.path)) {
+        const caretMap = new Map();
+        caretPathMap.set(rule.path, caretMap);
+      }
+
+      const workingPathMap = caretPathMap.get(rule.path);
+      if (!workingPathMap.has(element.base)) {
+        workingPathMap.set(element.base, 0);
+        if (element.brackets?.includes('+')) {
+          element.brackets[element.brackets.indexOf('+')] = '0';
+        }
+      } else {
+        element.brackets?.forEach((bracket: string, index: number) => {
+          if (bracket === '+') {
+            const newIndex = workingPathMap.get(element.base) + 1;
+            element.brackets[index] = newIndex.toString();
+            workingPathMap.set(element.base, newIndex);
+          } else if (bracket === '=') {
+            const currentIndex = workingPathMap.get(element.base);
+            element.brackets[index] = currentIndex.toString();
+          }
+        });
+      }
+    });
```

Comment:
Is this code exactly the same as lines 72-89, except it's working on `workingPathMap` instead of `pathMap`?  If so, can we extract it into a separate private function and just call it from both places (passing in the respective map as appropriate)?